### PR TITLE
Switch test coverage collection to gcovr

### PR DESCRIPTION
**Pull request by KochynVolodymyr**
### Motivate of the pull request
 - [ ] To address an existing issue. If so, please provide a link to the issue: <issue id>
 - [x] Breaking new feature. If so, please describe details in the description part.

 ### Describe the technical details
switch to gcovr for test coverage. 
gcovr>=6.0 is required
image

 
 ### Which part of the code base require a change
 
 - [ ] Library: <Specify the library name>
 - [ ] Plug-in: <Specify the plugin name>
 - [ ] Engine
 - [ ] Documentation
 - [ ] Regression tests
 - [x] Continous Integration (CI) scripts

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -72,16 +72,20 @@ test/unittest-d: run-cmake-debug
 	cmake --build dbuild --target unittest -j $(CPU_CORES)
 	pushd dbuild && $(XVFB) tests/unittest/unittest && popd
 
-test/unittest-coverage: run-cmake-coverage
-	cmake --build coverage-build --target unittest -j $(CPU_CORES)
-	pushd coverage-build && $(XVFB) tests/unittest/unittest && popd
 
-coverage-build/foedag.coverage: test/unittest-coverage
-	lcov --no-external --exclude "*_test.cpp" --capture --directory coverage-build/CMakeFiles/foedag.dir --base-directory src --output-file coverage-build/foedag.coverage
+# test/unittest-coverage: run-cmake-coverage
```
Comment (KochynVolodymyr) [inline Makefile:76]: Why we need this comments?

Comment (shivaahir1508) [reply Makefile:76]: it is an old code, just commented out for now, will delete soon.

**Review by ravic-rs — CHANGES_REQUESTED** [review]
@shivaahir1508 assigning this to you to fix the gcovr version in Ubuntu.

File: .github/workflows/install_ubuntu_dependencies_build.sh [hunk 2]
```diff
@@ -13,6 +14,7 @@ sudo apt install -y \
   xorg \
   qtbase5-dev qtchooser qt5-qmake qtbase5-dev-tools \
   xvfb \
+#  gcovr \
```
Comment (KochynVolodymyr) [inline .github/workflows/install_ubuntu_dependencies_build.sh:17]: Why we need this comment?

Comment (shivaahir1508) [reply .github/workflows/install_ubuntu_dependencies_build.sh]: it is removed.

File: .github/workflows/install_ubuntu_dependencies_build.sh [hunk 3]
```diff
@@ -1,4 +1,5 @@
 # Install required dependencies for Ubuntu systems
+pip install gcovr==6.0
```
Comment (KochynVolodymyr) [inline .github/workflows/install_ubuntu_dependencies_build.sh:2]: @shivaahir1508 since we have this, maybe we don't need lcov \ in dependencies. Please check that.

Comment (ravic-rs) [reply .github/workflows/install_ubuntu_dependencies_build.sh:2]: @shivaahir1508 shouldn't this be pip3 ?

Comment (shivaahir1508) [reply .github/workflows/install_ubuntu_dependencies_build.sh:2]: Hi, I will do the changes.

File: .github/workflows/install_centos_dependencies_build.sh [hunk 4]
```diff
@@ -1,17 +1,24 @@
+set -e
 # Install required dependencies for CentOS systems
 yum update -y
 yum group install -y "Development Tools" 
 yum install -y epel-release 
 curl -C - -O https://cmake.org/files/v3.15/cmake-3.15.7-Linux-x86_64.tar.gz
 tar xzf cmake-3.15.7-Linux-x86_64.tar.gz
 ln -s $PWD/cmake-3.15.7-Linux-x86_64/bin/cmake /usr/bin/cmake
+yum install -y openssh-server openssh-clients
```
Comment (KochynVolodymyr) [inline .github/workflows/install_centos_dependencies_build.sh:9]: @shivaahir1508 why this changes here? does it related to code coverage?

Comment (ravic-rs) [reply .github/workflows/install_centos_dependencies_build.sh:9]: @shivaahir1508 please create a separate PR for non-code coverage changes asap.

File: .github/workflows/install_centos_dependencies_build.sh [hunk 5]
```diff
@@ -24,6 +24,7 @@ yum install -y mesa-libGL-devel
 yum install -y libxcb libxcb-devel xcb-util xcb-util-devel libxkbcommon-devel libxkbcommon-x11-devel
 yum install -y xcb-util-image-devel xcb-util-keysyms-devel xcb-util-renderutil-devel xcb-util-wm-devel
 yum install -y gtk3-devel zip unzip
+
```
Comment (KochynVolodymyr) [inline .github/workflows/install_centos_dependencies_build.sh:27]: Why is it here?

File: .github/workflows/install_centos_dependencies_build.sh [hunk 6]
```diff
@@ -62,3 +63,4 @@ else
   cd ..
 fi
 yum clean all
+rm -rf qt-everywhere-src-5.15.4.tar.xz qt-everywhere-src-5.15.4
```
Comment (KochynVolodymyr) [inline .github/workflows/install_centos_dependencies_build.sh:66]: this is not related to code coverage

**Comment by nadeemyaseen-rs** [comment]
Closing this PR and then opening again as codecov is not picking the latest commit.

File: .github/workflows/install_centos_dependencies_build.sh [hunk 7]
```diff
@@ -61,4 +61,4 @@ else
   make install
   cd ..
 fi
-yum clean all
+yum clean all
```
Comment (KochynVolodymyr) [inline .github/workflows/install_centos_dependencies_build.sh:64]: @shivaahir1508 still exists

**Comment by nadeemyaseen-rs** [comment]
The coverage report is now available on [codecov](https://app.codecov.io/github/os-fpga/FOEDAG/tree/switch-coverage-to-gcovr)
On CI, the format of the coverage report will be XML and on the developer machine, it will be in HTML format as posted by KochynVolodymyr

**Comment by KochynVolodymyr** [comment]
@nadeemyaseen-rs two questions:
1. Why only branches coverage are displayed. There are also functions and lines.
2. Why branches coverage is different from the html?

**Comment by nadeemyaseen-rs** [comment]
> @nadeemyaseen-rs two questions:
> 
>     1. Why only branches coverage are displayed. There are also functions and lines.
> 
>     2. Why branches coverage is different from the html? I don't see 4.49% in the xml file.

4.49% is the coverage added due to this branch (git branch) changes. As per my current understanding, codecov platform only take into account the line coverage. Exploring its docs like [this one](https://docs.codecov.com/docs/coverage-percentages) may be helpful to understand the dashboard number.

**Review by ravic-rs — APPROVED** [review]
